contracts/bond/TimeLockMultiRewardBond.sol

```diff
@@ -0,0 +1,254 @@
+    }
+
+    /**
+     * @dev Must be called before the claimant debt tokens are burnt, as they are used with total supply in
+     *      calculating the rewards.
+     */
+    function _calculateRewardDebt(
+        address claimant,
+        uint256 claimantDebtTokens,
+        uint256 totalSupply
+    ) internal {
+        require(claimantDebtTokens < totalSupply, "Rewards: too much debt");
+
+        for (uint256 i = 0; i < EnumerableSetUpgradeable.length(_tokens); i++) {
+            address tokens = EnumerableSetUpgradeable.at(_tokens, i);
+            uint256 rewardDebt = (_rewardPool[tokens].amount *
+                claimantDebtTokens) / totalSupply;
+
+            _claimantToRewardPoolDebt[claimant][tokens] = rewardDebt;
+            emit RewardDebt(tokens, claimant, rewardDebt);
+        }
+    }
+
+    /**
+     * @notice Registers ERC20 tokens already transferred to the contract, adding them to the existing reward set.
+     *
+     * @dev Must be done before redemption is allowed.
+     *
+     * @param tokens ERC20 register of the tokens to transfer in and later distribute as rewards.
+     * @param amount total number of ERC20 tokens to use for reward.
+     * @param timeLock ms delay after redemption is allowed, only after which rewards can be claimed.
+     *        Value of zero means rewards can be claimed without delay after redemption is allowed.
+     *        Will overwrite the existing timeLock if there is already a reward from the ERC20 contract.
+     */
+    function _registerReward(
+        address tokens,
+        uint256 amount,
+        uint256 timeLock
+    ) internal whenNotPaused {
+        require(tokens != address(0), "Rewards: address is zero");
+        require(amount > 0, "Rewards: no reward amount");
+
+        emit RegisterReward(tokens, amount, timeLock);
+
+        _rewardPool[tokens].amount += amount;
+        _rewardPool[tokens].timeLock = timeLock;
+
+        require(
+            IERC20Upgradeable(tokens).balanceOf(address(this)) >=
+                _rewardPool[tokens].amount,
+            "Rewards: not enough held"
+        );
+    }
+
+    /**
+     * @notice Overwrites the existing time lock for rewards from a single ERC20.
+     *
+     * @param tokens ERC20 rewards already registered.
+     * @param timeLock ms to lock rewards after redemption is allowed.
+     */
+    function _updateRewardTimeLock(address tokens, uint256 timeLock)
+        internal
+        whenNotPaused
+    {
+        require(_hasRegisteredRewards(tokens), "Rewards: no reward tokens");
+
+        _rewardPool[tokens].timeLock = timeLock;
+
+        emit UpdateRewardTimeLock(tokens, timeLock);
+    }
+
+    /**
+     * @notice The time at which the debt tokens are redeemable.
+     *
+     * @dev Until a redemption time is set, no rewards are claimable.
+     */
+    function _setRedemptionTimestamp(uint256 timestamp) internal whenNotPaused {
+        require(timestamp != 0, "Rewards: zero is invalid time");
```

Comment:
do we want to require if this is in the future? `timestamp > block.timestamp`

Reply:
That would be a more sensible comparison ...wonder if Slither is going to complain about it 🤔

Reply:
https://docs.soliditylang.org/en/latest/units-and-global-variables.html?highlight=block.timestamp#block-and-transaction-properties

Solidity time is indeed in seconds 😶‍🌫️